mono/mini/mini-arm64.c

```diff
@@ -5114,6 +5114,66 @@ emit_store_regset_cfa (MonoCompile *cfg, guint8 *code, guint64 regs, int basereg
 	return code;
 }
 
+/* Same as emit_load_regset, but emit unwind info too */
+/* and emit the load in reverse order to satisfy win32 unwind info */
+/* CFA_OFFSET is the offset between the CFA and basereg */
+static __attribute__((__warn_unused_result__)) guint8*
+emit_load_regset_cfa (MonoCompile* cfg, guint8* code, guint64 stored_regs, guint64 regs, int basereg, int offset, int cfa_offset, guint64 no_cfa_regset) {
+
+	int i, j, pos, nregs;
+	guint32 cfa_regset = regs & ~no_cfa_regset;
+	guint32 cfa_stored_regset = stored_regs & ~no_cfa_regset;
+
+	pos = -1;
+
+	for (i = 0; i < 32; ++i) {
+		if (cfa_stored_regset & (1 << i)) {
+			pos += 1;
+		}
+	}
+
+	for (i = 32; i >= 0; --i) {
+		nregs = 1;
+		if (cfa_regset & (1 << i)) {
+			if (pos > 0 && pos % 2 && (cfa_regset & (1 << (i - 1))) && (i - 1 != ARMREG_SP)) {
+				if (offset < 256) {
+					arm_ldpx(code, i - 1, i, basereg, offset + (pos * 8) - 8);
+				}
+				else {
+					code = emit_ldrx(code, i, basereg, offset + (pos * 8));
+					code = emit_ldrx(code, i - 1, basereg, offset + (pos * 8) - 8);
+				}
+				nregs = 2;
+			}
+			else if (i == ARMREG_SP) {
+				arm_movspx(code, ARMREG_IP1, ARMREG_SP);
+				code = emit_ldrx(code, ARMREG_IP1, basereg, offset + (pos * 8));
+			}
+			else {
+				code = emit_ldrx(code, i, basereg, offset + (pos * 8));
+			}
+
+			if (mono_arch_unwindinfo_emit_epilog_codes(cfa_regset, cfa_stored_regset))
+			{
+				// if we loaded a different set of registers that we stored, we need to record what is actualy in the epilog
+				for (j = 0; j < nregs; ++j) {
+					if (cfa_regset & (1 << (i - j)))
+						mono_emit_unwind_op_restore_offset(cfg, code, i + j, (-cfa_offset) + offset + ((pos + j) * 8));
+				}
+			}
+
+			i -= nregs - 1;
```

Comment:
I'm not following this loop. Perhaps an interactive review makes sense. Otherwise, I would suggest some more comments and some asserts like `i >=0` (assuming that should be true).

Reply:
It's a confusing loop - it's mostly structured to match https://github.com/Unity-Technologies/mono/blob/unity-main/mono/mini/mini-arm64.c#L5089-L5122, otherwise I would have implemented it a bit differently.